### PR TITLE
copr: Add vector functions

### PR DESCRIPTION
**Pull request by EricZequan**
### What is changed and how it works?

Issue Number: ref #17290 
The PRE-PR of this PR: #17287 ,  do not merge this PR before the pre-pr megered.

What's Changed:
Implement these vector functions:

1. VEC_L1_DISTANCE
2. VEC_L2_DISTANCE
3. VEC_NEGATIVE_INNER_PRODUCT
4. VEC_COSINE_DISTANCE
5. VEC_L2_NORM

The implementation is aligned with [pgvector](https://github.com/pgvector/pgvector/blob/master/src/vector.c), [milvus](https://github.com/milvus-io/knowhere/blob/main/src/simd/distances_ref.cc) and [pgvecto.rs](https://github.com/tensorchord/pgvecto.rs/blob/main/crates/service/src/prelude/global/f32_cos.rs#L211), which uses float32 for intermediate results. The test cases are also [ported from pgvector](https://github.com/pgvector/pgvector/blob/master/test/expected/functions.out) to make sure that users will receive same results as pgvector.

This PR also rejects vectors containing NaNs or ±Inf. Currently TiDB allows such data, and this will be fixed in TiDB later PRs.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by EricZequan** [comment]
/cc @breezewish  
Is the way i push this pr right? plz take a review🙏

**Comment by breezewish** [comment]
@EricZequan You must not bring updates in origin/master into this PR. Otherwise the review cannot continue because there are too many unrelated changes and we don't know whether it is correct or not:

image

**Review by breezewish — APPROVED** [review]

File: components/tidb_query_expr/src/impl_vec.rs [hunk 1]
```diff
@@ -15,3 +15,271 @@ fn vec_as_text(a: VectorFloat32Ref, writer: BytesWriter) -> Result<BytesGuard> {
...
+                    "expected {:?}, got {:?}",
+                    expected_output,
+                    output
+                );
+            }
+        }
+    }
+
+    #[test]
+    fn test_l2_distance() {
```
Comment (JaySon-Huang) [inline components/tidb_query_expr/src/impl_vec.rs:127]: Add some case that defining the (l1/l2/negative_inner_product/cosine_distance) distance between vector data and "NULL"?

Comment (EricZequan) [reply components/tidb_query_expr/src/impl_vec.rs:127]: i add this sample in the code to test (l1/l2/negative_inner_product/cosine_distance):
```rust
let err_cases = vec![(vec![1.0, 2.0], vec![3.0]), (vec![1.0, 2.0], vec![])];
        for (arg1, arg2) in err_cases {
            let arg1 = VectorFloat32::new(arg1).unwrap();
            let arg2 = VectorFloat32::new(arg2).unwrap();
            let output: Result<Option<Real>> = RpnFnScalarEvaluator::new()
                .push_param(arg1)
                .push_param(arg2)
                .evaluate(ScalarFuncSig::VecNegativeInnerProductSig);
            assert!(output.is_err(), "expected error, got {:?}", output);
        }
```

Comment (breezewish) [reply components/tidb_query_expr/src/impl_vec.rs:127]: This only tests error output, could you add test case for NULL inputs?

Comment (EricZequan) [reply components/tidb_query_expr/src/impl_vec.rs:129]: modify into :
```rust
let ok_cases = vec![
            (Some(vec![0.0, 0.0]), Some(vec![3.0, 4.0]), Some(5.0)),
            (Some(vec![0.0, 0.0]), Some(vec![0.0, 1.0]), Some(1.0)),
            (Some(vec![3e38]), Some(vec![-3e38]), Some(f64::INFINITY)),
            (Some(vec![1.0, 2.0]), None, None),
        ];
        for (arg1, arg2, expected_output) in ok_cases {
            let arg1 = arg1.map(|v| VectorFloat32::new(v).unwrap());
            let arg2 = arg2.map(|v| VectorFloat32::new(v).unwrap());
            let output: Option<Real> = RpnFnScalarEvaluator::new()
                .push_param(arg1)
                .push_param(arg2)
                .evaluate(ScalarFuncSig::VecL2DistanceSig)
                .unwrap();
            if expected_output.is_none() {
                assert!(output.is_none(), "expected None, got {:?}", output);
            } else {
                assert!(
                    option_real_eq(output, expected_output.map(|x| Real::new(x).unwrap())),
                    "expected {:?}, got {:?}",
                    expected_output,
                    output
                );
            }
        }
```

**Comment by breezewish** [comment]
Please fix the lint.

**Review by breezewish — APPROVED** [review]
Please fix the lint. The rest LGTM.

**Review by gengliqi — APPROVED** [review]
